### PR TITLE
Memory Management Milestone Proposal

### PR DESCRIPTION
**Pull request by LeshengJin**

### CODE REVIEW

**Review by GV1028 — COMMENTED** [review]
Good work! I think the testref_count is a really neat idea!

How does your design handle cases like an array of lists? For example, if the lists group wants to use your interfaces to implement an array of lists, can they do so?

Does your design choices also account for defragmentation? Do you need any extra utility functions to do that?

**Comment by mb64** [comment]
Thanks for the review!

> if the lists group wants to use your interfaces to implement an array of lists, can they do so?

Lists, strings, and possibly bigints are the only special dynamic layouts, and they will need special handling in the `codeGenAlloc` function.  AFAICT there will only be two kinds of special dynamic layouts here:
 - Lists of pointers, which are dynamically sized and all pointers
 - Big chunks of bytes, dynamically sized with no pointers in them

But we'll need to coordinate with these groups to work out exactly what the layouts are. Hopefully that could happen tomorrow during class (?)

> Does your design choices also account for defragmentation? Do you need any extra utility functions to do that?

We're intentionally avoiding a shadow stack in this design, and without precise stack access any kind of moving GC is fundamentally impossible.  So excepting very fancy things like semi-conservative collection techniques or UMass's Mesh (which isn't implementable in WASM) we'll only be able to defragment as much as regular Python and regular `malloc`/`free`.

And anyway this is the concern of next week, when we implement the allocator, for this week we're just using a bump pointer and counting references :smile:

File: compiler.ts [hunk 1]
```diff
@@ -25,24 +25,28 @@ type CompileResult = {
...
-    localDefines.push(`(local $${v} i32)`);
+  localDefines.push(`(local $$last i32)`);
+  locals.forEach((type, name) => {
+    localDefines.push(`(local $${name} i32)`);
   });
   return localDefines;
 }
 
 export function compile(ast: Program<[Type, SourceLocation]>, env: GlobalEnv) : CompileResult {
+  console.log(ast);
```
Comment (KamranAlipour) [inline compiler.ts:38]: please make sure you clean up all stray code like these console.log s.

**Review by KamranAlipour — COMMENTED** [review]
Great progress, overall looks good, I think the main hurdle right now is to think of inheritance for GC

**Review by KamranAlipour — COMMENTED** [review]
Great progress, overall looks good, I think the main hurdle right now is to think of inheritance for GC

File: compiler.ts (hunk 1 above)
Comment (KamranAlipour) [inline compiler.ts:38]: please make sure you clean up all stray code like these console.log s.

**Pull request closed without merging**